src/main/java/com/cloudant/client/api/Database.java

```diff
@@ -1029,135 +1029,116 @@ public void ensureFullCommit() {
         try {
-            client.getGson().fromJson(selectorJson, JsonObject.class);
+            selectorObject = getGson().fromJson(selectorJson, JsonObject.class);
         } catch (JsonParseException e) {
             isObject = false;
         }
 
-        if (!isObject) {
-            // needs to start with selector
-            if (!(selectorJson.trim().startsWith("\"selector\""))) {
+        if(!isObject) {
+            if(selectorJson.startsWith("\"selector\"")) {
+                selectorJson = selectorJson.substring(selectorJson.indexOf(":") + 1,
+                        selectorJson.length()).trim();
+                selectorObject = getGson().fromJson(selectorJson, JsonObject.class);
+            } else {
                 throw new JsonParseException("selectorJson should be valid json or like " +
                         "\"selector\": {...} ");
             }
         }
 
-        StringBuilder finalbody = new StringBuilder();
-        if (isObject) {
-            finalbody.append("{\"selector\": ")
-                    .append(selectorJson);
+        if(selectorObject.has("selector")) {
+            indexObject.add("selector", selectorObject.get("selector"));
         } else {
-            //old support
-            finalbody.append("{" + selectorJson);
+            indexObject.add("selector", selectorObject);
         }
 
-        if (rf != null) {
-            finalbody.append(",")
-                    .append(rf.toString());
+        if(fieldsArray.size() > 0) {
+            indexObject.add("fields", fieldsArray);
         }
-        if (so != null) {
-            finalbody.append(",")
-                    .append(so.toString());
+        if(sortArray.size() > 0) {
+            indexObject.add("sort", sortArray);
         }
         if (options.getLimit() != null) {
-            finalbody.append(",")
-                    .append("\"limit\": ")
-                    .append(options.getLimit());
+            indexObject.addProperty("limit", options.getLimit());
         }
         if (options.getSkip() != null) {
-            finalbody.append(",")
-                    .append("\"skip\": ")
-                    .append(options.getSkip());
+            indexObject.addProperty("skip", options.getSkip());
         }
         if (options.getReadQuorum() != null) {
-            finalbody.append(",")
-                    .append("\"r\": ")
-                    .append(options.getReadQuorum());
+            indexObject.addProperty("r", options.getReadQuorum());
         }
         if (options.getUseIndex() != null) {
-            finalbody.append(",")
-                    .append("\"use_index\": ")
-                    .append(options.getUseIndex());
+            JsonArray useIndexArray = new JsonArray();
+            JsonPrimitive element = new JsonPrimitive(options.getUseIndex());
+            useIndexArray.add(element);
+            indexObject.add("use_index", getGson().fromJson(options.getUseIndex(), JsonElement.class));
```

Comment:
I think this is the only line needed, but it might be worth changing `JsonElement.class` to `JsonArray.class` since we should probably break if the `getUseIndex()` String doesn't turn into an array.

I think this also highlights a place in `FindByIndexOptions` that is still doing String JSON operations. That is creating the `useIndex. Probably worth changing the internal type for `useIndex` in `FindByIndexOptions` into a `JsonArray` and `return useIndex.toString();` from the `getUseIndex()` method. Although the `JsonArray` -> `String` -> `JsonArray` is slightly cumbersome it avoids exposing the internal JSON choice on the public `getUseIndex()` method. An alternative might be to use a reduced visibility method that returns the `JsonArray` directly to avoid the journey via String.